File: internal/recur/recur.go
// Package recur handles task recurrence: a canonical RRULE subset with just
// enough grammar for a personal tracker. Parse validates a rule and Next
// computes the following occurrence, so the server can roll a completed
// recurring task forward. FromNatural maps everyday phrases ("every weekday",
// "2 weeks") onto the same canonical form the web frontend mirrors.
//
// The subset is deliberately small — anything outside it is rejected:
//
//	FREQ=DAILY|WEEKLY|MONTHLY|YEARLY   (required)
//	INTERVAL=n                          (optional, n>=1)
//	BYDAY=MO,TU,WE,TH,FR,SA,SU          (optional, WEEKLY only)
//
// Expand is the exception: it delegates to a full iCalendar RRULE engine for
// the ics syncer's windowed feed expansion, which needs COUNT/UNTIL/etc. that
// task recurrence never uses.
package recur

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Freq is a recurrence frequency.
type Freq int

const (
	Daily Freq = iota
	Weekly
	Monthly
	Yearly
)

// Rule is a parsed, validated recurrence.
type Rule struct {
	Freq     Freq
	Interval int            // always >= 1
	ByDay    []time.Weekday // WEEKLY only; sorted MO..SU; empty means every week
}

// weekdayCodes maps the canonical two-letter BYDAY codes to weekdays.
var weekdayCodes = map[string]time.Weekday{
	"MO": time.Monday,
	"TU": time.Tuesday,
	"WE": time.Wednesday,
	"TH": time.Thursday,
	"FR": time.Friday,
	"SA": time.Saturday,
	"SU": time.Sunday,
}

// codeOf is the inverse of weekdayCodes.
var codeOf = map[time.Weekday]string{
	time.Monday:    "MO",
	time.Tuesday:   "TU",
	time.Wednesday: "WE",
	time.Thursday:  "TH",
	time.Friday:    "FR",
	time.Saturday:  "SA",
	time.Sunday:    "SU",
}

var freqNames = map[Freq]string{
	Daily:   "DAILY",
	Weekly:  "WEEKLY",
	Monthly: "MONTHLY",
	Yearly:  "YEARLY",
}

// isoIndex orders weekdays Monday=0 .. Sunday=6 for week arithmetic and BYDAY
// sorting (Go's time.Weekday puts Sunday first).
func isoIndex(d time.Weekday) int { return (int(d) + 6) % 7 }

// Parse validates a canonical RRULE-subset string. Keys are case-insensitive;
// unknown keys, duplicate keys, unsupported values, and BYDAY on a non-weekly
// rule are all errors.
func Parse(rule string) (Rule, error) {
	rule = strings.TrimSpace(rule)
	if rule == "" {
		return Rule{}, fmt.Errorf("recurrence is empty")
	}
	var (
		r        Rule
		haveFreq bool
		seen     = map[string]bool{}
	)
	r.Interval = 1
	for _, part := range strings.Split(rule, ";") {
		key, val, ok := strings.Cut(part, "=")
		if !ok {
			return Rule{}, fmt.Errorf("recurrence %q: %q is not KEY=VALUE", rule, part)
		}
		key = strings.ToUpper(strings.TrimSpace(key))
		val = strings.TrimSpace(val)
		if seen[key] {
			return Rule{}, fmt.Errorf("recurrence %q: duplicate %s", rule, key)
		}
		seen[key] = true
		switch key {
		case "FREQ":
			f, ok := parseFreq(val)
			if !ok {
				return Rule{}, fmt.Errorf("recurrence %q: FREQ=%s is not one of DAILY, WEEKLY, MONTHLY, YEARLY", rule, val)
			}
			r.Freq, haveFreq = f, true
		case "INTERVAL":
			n, ok := atoiStrict(val)
			if !ok || n < 1 {
				return Rule{}, fmt.Errorf("recurrence %q: INTERVAL must be an integer >= 1, got %q", rule, val)
			}
			r.Interval = n
		case "BYDAY":
			days, err := parseByDay(val)
			if err != nil {
				return Rule{}, fmt.Errorf("recurrence %q: %w", rule, err)
			}
			r.ByDay = days
		default:
			return Rule{}, fmt.Errorf("recurrence %q: unsupported key %s", rule, key)
		}
	}
	if !haveFreq {
		return Rule{}, fmt.Errorf("recurrence %q: FREQ is required", rule)
	}
	if len(r.ByDay) > 0 && r.Freq != Weekly {
		return Rule{}, fmt.Errorf("recurrence %q: BYDAY is only allowed with FREQ=WEEKLY", rule)
	}
	return r, nil
}

// atoiStrict parses s as a base-10 non-negative integer, accepting ASCII digits
// only. Unlike strconv.Atoi it rejects a leading sign, so "+2" and "-2" are not
// valid intervals (parity with the web parser's digits-only rule).
func atoiStrict(s string) (int, bool) {
	if s == "" {
		return 0, false
	}
	for _, c := range s {
		if c < '0' || c > '9' {
			return 0, false
		}
	}
	n, err := strconv.Atoi(s)
	if err != nil {
		return 0, false
	}
	return n, true
}

func parseFreq(s string) (Freq, bool) {
	switch strings.ToUpper(s) {
	case "DAILY":
		return Daily, true
	case "WEEKLY":
		return Weekly, true
	case "MONTHLY":
		return Monthly, true
	case "YEARLY":
		return Yearly, true
	}
	return 0, false
}

// parseByDay parses a comma list of BYDAY codes into sorted, deduplicated
// weekdays.
func parseByDay(s string) ([]time.Weekday, error) {
	seen := map[time.Weekday]bool{}
	var days []time.Weekday
	for _, code := range strings.Split(s, ",") {
		code = strings.ToUpper(strings.TrimSpace(code))
		wd, ok := weekdayCodes[code]
		if !ok {
			return nil, fmt.Errorf("BYDAY %q is not one of MO,TU,WE,TH,FR,SA,SU", code)
		}
		if !seen[wd] {
			seen[wd] = true
			days = append(days, wd)
		}
	}
	if len(days) == 0 {
		return nil, fmt.Errorf("BYDAY must list at least one day")
	}
	sortWeekdays(days)
	return days, nil
}

func sortWeekdays(days []time.Weekday) {
	// Small fixed set: insertion sort by ISO index keeps MO..SU order.
	for i := 1; i < len(days); i++ {
		for j := i; j > 0 && isoIndex(days[j]) < isoIndex(days[j-1]); j-- {
			days[j], days[j-1] = days[j-1], days[j]
		}
	}
}

// String renders the rule back to its canonical form (INTERVAL omitted when 1,
// BYDAY listed in MO..SU order). Parse(r.String()) round-trips.
func (r Rule) String() string {
	var b strings.Builder
	b.WriteString("FREQ=")
	b.WriteString(freqNames[r.Freq])
	if r.Interval > 1 {
		b.WriteString(";INTERVAL=")
		b.WriteString(strconv.Itoa(r.Interval))
	}
	if len(r.ByDay) > 0 {
		codes := make([]string, len(r.ByDay))
		for i, d := range r.ByDay {
			codes[i] = codeOf[d]
		}
		b.WriteString(";BYDAY=")
		b.WriteString(strings.Join(codes, ","))
	}
	return b.String()
}

// Next returns the first occurrence strictly after t, keeping t's local
// wall-clock time of day (and thus its location — recurrence preserves local
// wall-clock time, so the caller's timezone is authoritative and the absolute
// instant shifts across DST). Monthly and yearly advance by whole months and
// clamp the day to the target month's length (Jan 31 -> Feb 28), matching how a
// calendar rolls a month-day that does not exist forward.
func (r Rule) Next(t time.Time) time.Time {
	// A validated Rule always has Interval >= 1, but guard a hand-built one so a
	// zero/negative interval can never produce a non-advancing loop in callers.
	if r.Interval < 1 {
		r.Interval = 1
	}
	switch r.Freq {
	case Weekly:
		if len(r.ByDay) > 0 {
			return r.nextWeeklyByDay(t)
		}
		return t.AddDate(0, 0, 7*r.Interval)
	case Monthly:
		return AddMonthsClamped(t, r.Interval)
	case Yearly:
		return AddMonthsClamped(t, 12*r.Interval)
	default: // Daily
		return t.AddDate(0, 0, r.Interval)
	}
}

// nextWeeklyByDay finds the next BYDAY weekday after t. Within t's own week
// (Monday-anchored) it stays in-week; once the week's listed days are
// exhausted it jumps Interval weeks forward to the first listed day.
func (r Rule) nextWeeklyByDay(t time.Time) time.Time {
	cur := isoIndex(t.Weekday())
	for _, d := range r.ByDay {
		if bi := isoIndex(d); bi > cur {
			return t.AddDate(0, 0, bi-cur)
		}
	}
	// No later day this week: advance to the next active week's first day.
	daysToNextMonday := 7 - cur
	weekJump := daysToNextMonday + (r.Interval-1)*7
	return t.AddDate(0, 0, weekJump+isoIndex(r.ByDay[0]))
}

// AddMonthsClamped adds months to t, clamping the day to the last day of the
// target month so no overflow into the following month occurs (Jan 31 + 1 month
// = Feb 28, not Mar 3) and preserving t's clock and location. Exported so CLI
// month/year offsets clamp identically to recurrence.
func AddMonthsClamped(t time.Time, months int) time.Time {
	y, m, d := t.Date()
	total := int(m) - 1 + months
	ty := y + total/12
	tm := time.Month(total%12) + 1
	if dim := daysInMonth(ty, tm); d > dim {
		d = dim
	}
	return time.Date(ty, tm, d, t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), t.Location())
}

// daysInMonth returns the number of days in the given month (day 0 of the next
// month is the last day of this one).
func daysInMonth(year int, month time.Month) int {
	return time.Date(year, month+1, 0, 0, 0, 0, 0, time.UTC).Day()
}

// Humanize renders a canonical rule as a short English phrase, or returns the
// rule verbatim if it does not parse.
func Humanize(rule string) string {
	r, err := Parse(rule)
	if err != nil {
		return rule
	}
	unit := map[Freq]string{Daily: "day", Weekly: "week", Monthly: "month", Yearly: "year"}[r.Freq]
	if r.Freq == Weekly && len(r.ByDay) > 0 {
		if isWeekdaySet(r.ByDay) {
			return "every weekday"
		}
		// Three-letter abbreviations ("Mon, Wed, Fri"), matching the web UI.
		names := make([]string, len(r.ByDay))
		for i, d := range r.ByDay {
			names[i] = d.String()[:3]
		}
		return "every " + strings.Join(names, ", ")
	}
	if r.Interval == 1 {
		return "every " + unit
	}
	return fmt.Sprintf("every %d %ss", r.Interval, unit)
}

// isWeekdaySet reports whether days is exactly Monday through Friday.
func isWeekdaySet(days []time.Weekday) bool {
	if len(days) != 5 {
		return false
	}
	want := []time.Weekday{time.Monday, time.Tuesday, time.Wednesday, time.Thursday, time.Friday}
	for i, d := range days {
		if d != want[i] {
			return false
		}
	}
	return true
}

File: web/src/lib/recur.ts
// A TypeScript mirror of the Go internal/recur package: the canonical RRULE
// subset a task's `recurrence` field carries, plus the natural-language grammar
// that maps everyday phrases onto it. Kept in lockstep with
// internal/recur/natural.go and recur.go — the same phrase yields the same
// canonical string on both sides, so a rule typed in the web UI reads back the
// same way the CLI wrote it.
//
// The subset (anything outside it is rejected):
//   FREQ=DAILY|WEEKLY|MONTHLY|YEARLY   (required)
//   INTERVAL=n                          (optional, n>=1; omitted when 1)
//   BYDAY=MO,TU,WE,TH,FR,SA,SU          (optional, WEEKLY only; MO..SU order)

export type Freq = "DAILY" | "WEEKLY" | "MONTHLY" | "YEARLY";

export interface Rule {
  freq: Freq;
  /** Always >= 1. */
  interval: number;
  /** BYDAY codes, sorted MO..SU; empty means every occurrence. */
  byDay: string[];
}

// ISO weekday order Monday..Sunday, for BYDAY sorting and canonical output.
const CODES = ["MO", "TU", "WE", "TH", "FR", "SA", "SU"] as const;
const CODE_INDEX: Record<string, number> = Object.fromEntries(CODES.map((c, i) => [c, i]));

// Full weekday names (Monday-first) paired with their BYDAY codes, for the
// natural-language weekday-list parser.
const WEEKDAY_NAMES: [string, string][] = [
  ["monday", "MO"],
  ["tuesday", "TU"],
  ["wednesday", "WE"],
  ["thursday", "TH"],
  ["friday", "FR"],
  ["saturday", "SA"],
  ["sunday", "SU"],
];

const FREQS: readonly string[] = ["DAILY", "WEEKLY", "MONTHLY", "YEARLY"];

const UNIT_FREQ: Record<string, Freq> = {
  day: "DAILY",
  week: "WEEKLY",
  month: "MONTHLY",
  year: "YEARLY",
};

const UNIT_NAME: Record<Freq, string> = {
  DAILY: "day",
  WEEKLY: "week",
  MONTHLY: "month",
  YEARLY: "year",
};

const ABBREV: Record<string, string> = {
  MO: "Mon",
  TU: "Tue",
  WE: "Wed",
  TH: "Thu",
  FR: "Fri",
  SA: "Sat",
  SU: "Sun",
};

/** Render a rule to canonical form: INTERVAL omitted when 1, BYDAY in MO..SU. */
function ruleString(r: Rule): string {
  let out = `FREQ=${r.freq}`;
  if (r.interval > 1) out += `;INTERVAL=${r.interval}`;
  if (r.byDay.length > 0) out += `;BYDAY=${r.byDay.join(",")}`;
  return out;
}

/** Validate a canonical RRULE-subset string. Case-insensitive keys; INTERVAL
 *  must be an integer >= 1; BYDAY is WEEKLY-only; duplicate or unknown keys are
 *  errors. Returns the parsed rule, or null when malformed. Mirror of Go Parse. */
export function parseRule(rule: string): Rule | null {
  const trimmed = rule.trim();
  if (trimmed === "") return null;
  let freq: Freq | null = null;
  let interval = 1;
  let byDay: string[] = [];
  const seen = new Set<string>();
  for (const part of trimmed.split(";")) {
    const eq = part.indexOf("=");
    if (eq < 0) return null;
    const key = part.slice(0, eq).trim().toUpperCase();
    const val = part.slice(eq + 1).trim();
    if (seen.has(key)) return null;
    seen.add(key);
    switch (key) {
      case "FREQ": {
        const f = val.toUpperCase();
        if (!FREQS.includes(f)) return null;
        freq = f as Freq;
        break;
      }
      case "INTERVAL": {
        if (!/^\d+$/.test(val)) return null;
        const n = parseInt(val, 10);
        if (n < 1) return null;
        interval = n;
        break;
      }
      case "BYDAY": {
        const days = parseByDay(val);
        if (!days) return null;
        byDay = days;
        break;
      }
      default:
        return null;
    }
  }
  if (freq === null) return null;
  if (byDay.length > 0 && freq !== "WEEKLY") return null;
  return { freq, interval, byDay };
}

function parseByDay(s: string): string[] | null {
  const seen = new Set<string>();
  const days: string[] = [];
  for (const raw of s.split(",")) {
    const code = raw.trim().toUpperCase();
    if (!(code in CODE_INDEX)) return null;
    if (!seen.has(code)) {
      seen.add(code);
      days.push(code);
    }
  }
  if (days.length === 0) return null;
  days.sort((a, b) => CODE_INDEX[a] - CODE_INDEX[b]);
  return days;
}

/** Map an everyday phrase to a canonical RRULE, or null if unrecognized. It is
 *  case-insensitive and a leading "every " is optional throughout. Mirror of Go
 *  FromNatural — grammar:
 *    "every day" / "daily"                 -> FREQ=DAILY
 *    "every week" / "weekly"               -> FREQ=WEEKLY
 *    "every month" / "monthly"             -> FREQ=MONTHLY
 *    "every year" / "yearly"               -> FREQ=YEARLY
 *    "every weekday"                       -> FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR
 *    "every N days|weeks|months|years"     -> FREQ=...;INTERVAL=N
 *    "every mon,wed" / "monday, friday"    -> FREQ=WEEKLY;BYDAY=...
 *  Weekday names may be full or any prefix of at least three letters. */
export function fromNatural(s: string): string | null {
  const orig = s.trim();
  if (orig === "") return null;
  let body = orig.toLowerCase();
  if (body.startsWith("every ")) body = body.slice("every ".length).trim();
  if (body === "") return null;

  switch (body) {
    case "day":
    case "daily":
      return "FREQ=DAILY";
    case "week":
    case "weekly":
      return "FREQ=WEEKLY";
    case "month":
    case "monthly":
      return "FREQ=MONTHLY";
    case "year":
    case "yearly":
      return "FREQ=YEARLY";
    case "weekday":
    case "weekdays":
      return "FREQ=WEEKLY;BYDAY=MO,TU,WE,TH,FR";
  }

  // "N days" / "N weeks" / "N months" / "N years".
  const fields = body.split(/\s+/);
  if (fields.length === 2 && /^\d+$/.test(fields[0])) {
    const freq = UNIT_FREQ[fields[1].replace(/s$/, "")];
    if (freq) {
      const n = parseInt(fields[0], 10);
      if (n < 1) return null;
      return ruleString({ freq, interval: n, byDay: [] });
    }
  }

  // A weekday list: "mon,wed,fri" or "monday, friday".
  const days = parseWeekdayList(body);
  if (days) return ruleString({ freq: "WEEKLY", interval: 1, byDay: days });

  return null;
}

function parseWeekdayList(s: string): string[] | null {
  const seen = new Set<string>();
  const days: string[] = [];
  for (const raw of s.split(",")) {
    const code = parseWeekdayName(raw.trim());
    if (!code) return null;
    if (!seen.has(code)) {
      seen.add(code);
      days.push(code);
    }
  }
  if (days.length === 0) return null;
  days.sort((a, b) => CODE_INDEX[a] - CODE_INDEX[b]);
  return days;
}

/** Resolve a full weekday name or a prefix of at least three letters to its
 *  BYDAY code (three letters disambiguates all seven days). */
function parseWeekdayName(tok: string): string | null {
  if (tok.length < 3) return null;
  let match: string | null = null;
  let found = 0;
  for (const [name, code] of WEEKDAY_NAMES) {
    if (name.startsWith(tok)) {
      match = code;
      found++;
    }
  }
  return found === 1 ? match : null;
}

/** Render a canonical rule as short, friendly English ("every 2 weeks",
 *  "every Mon, Wed"), or the rule verbatim if it does not parse. Weekday lists
 *  use three-letter abbreviations; a full Mon–Fri set reads "every weekday". */
export function humanize(rule: string): string {
  const r = parseRule(rule);
  if (!r) return rule;
  if (r.freq === "WEEKLY" && r.byDay.length > 0) {
    if (isWeekdaySet(r.byDay)) return "every weekday";
    return "every " + r.byDay.map((c) => ABBREV[c]).join(", ");
  }
  if (r.interval === 1) return `every ${UNIT_NAME[r.freq]}`;
  return `every ${r.interval} ${UNIT_NAME[r.freq]}s`;
}

function isWeekdaySet(byDay: string[]): boolean {
  const want = ["MO", "TU", "WE", "TH", "FR"];
  return byDay.length === want.length && want.every((c, i) => byDay[i] === c);
}
